### PR TITLE
Issues 277 and 242 - exceptions entropy transformer

### LINKED ISSUE #277: Introduce EntropyTransformer and EPR strategy
Implement the `EntropyTransformer` component and its single per-strategy reduction function.

## Design: one reduction function per strategy
A strategy is a **single function** `reduction(s_kj, axis)` that pools NaN-aware entropy contributions over a token axis into the feature vector. The transformer calls the *same* function for both methods — only the token axis changes:

```python
# self.reduction_fn is a property that resolves self.reduction to the function (stateless, no fit needed)
def transform(self, X):                 # sequence level
    s_kj = self.entropy_contributions(X, self.k)        # (n, T, k)
    return self.reduction_fn(s_kj, axis=1)                 # pool ALL tokens → (n, n_features)

def transform_tokens(self, X):          # token level
    s_kj = self.entropy_contributions(X, self.k)        # (n, T, k)
    windows = np.expand_dims(s_kj, axis=2)              # (n, T, 1, k) — one 1-token window per token
    return self.reduction_fn(windows, axis=2)             # pool each window → (n, T, n_features)
```

Because pooling a **1-token window** with `nanmean`/`nanmax` is the identity, the token-level features are exactly what the sequence pools — so token and sequence features come from the *same* function and **cannot drift apart**. (This is what fixes the WEPR token/sequence mismatch at the root: WEPR's both-branch token features fall out automatically.)

### `STRATEGIES` — how EPR and WEPR each use one function
```python
def _epr(x, axis):
    is_nan = np.isnan(x)
    padded = np.all(is_nan, axis=-1, keepdims=True)   # fully-NaN (padded) tokens
    s = np.nansum(x, axis=-1, keepdims=True)          # sum over k (rank axis)
    s = np.where(padded, np.nan, s)                   # nansum gave 0 for padded tokens → restore NaN
    return np.nanmean(s, axis=axis)                   # pool over the token axis

def _wepr(x, axis):                                   # #242
    mean_branch = np.nanmean(x, axis=axis)
    max_branch = np.nanmax(x, axis=axis)
    return np.concatenate([mean_branch, max_branch], axis=-1)

STRATEGIES = {"epr": _epr, "wepr": _wepr}
```

| `reduction(x, axis)` | token level (1-token window, `axis=2`) | sequence level (all tokens, `axis=1`) | `n_features` |
|---|---|---|---|
| **`_epr`** | `nansum_k(s_kj)` → `(n, T, 1)` | `nanmean_T(nansum_k(s_kj))` → `(n, 1)` | `1` |
| **`_wepr`** | `[s_kj, s_kj]` (both branches) → `(n, T, 2k)` | `[nanmean_T(s_kj), nanmax_T(s_kj)]` → `(n, 2k)` | `2k` |

For WEPR, the classifier's `coef_ = [mean_weights, max_weights]` then applies **both** branches per token (token logit `s_kj @ (mean_weights + max_weights)`) and at the sequence level reproduces the legacy `mean_term + max_term`.

```mermaid
flowchart LR
    A["s_kj : (n, T, k)<br/>NaN-aware (#239)"]
    A -->|"reduction(s_kj, axis=1)<br/>pool ALL tokens"| SEQ["transform<br/>(n, n_features)"]
    A -->|"reduction(s_kj[:, :, None, :], axis=2)<br/>pool 1-token windows"| TOK["transform_tokens<br/>(n, T, n_features)"]
```

## Input contract
`X` is `LogProbParser`'s output: a NaN-padded `(n, T, k)` float array (real positions finite, padding `NaN`). The transformer calls `self.entropy_contributions(X, self.k)` (#239) to get `s_kj`, then applies the single `reduction` function with NaN-aware ops.

## NaN conventions (must hold for every strategy)
- **Missing ranks** within a real token: `NaN` along the `k` axis → `nansum`/`nanmean` over ranks treat them as skipped (equivalent to legacy zero-padding); `nanmean`/`nanmax` over tokens exclude them.
- **Padded tokens** (fully-`NaN`, `np.all(np.isnan(x), axis=-1)`): must be excluded from the token-axis pooling. Pure `nanmean`/`nanmax` pooling (WEPR) excludes them automatically. **Any rank-axis `nansum` (EPR) turns an all-`NaN` token into `0`, not `NaN`**, so restore `NaN` for fully-padded tokens before the token pooling — see the `np.where` in `_epr`. This reproduces the legacy "average over real tokens" math and keeps calibration weights valid.
- **Empty sequences** (zero real tokens → the *whole* reduced feature row is `NaN`): in `transform` (sequence path) only, **emit an `EmptySequenceWarning`** (degenerate input shouldn't pass silently) and map the all-`NaN` feature row to a **zero vector**, so the inherited `Pipeline.predict_proba` returns the calibrated baseline `sigmoid(intercept_)` (matches legacy; required because #240 does not override `predict_proba` and `LogisticRegression` rejects `NaN`). Detection is exact — a non-empty sequence always yields finite features, so a genuinely zero-entropy sequence reads as `0.0`, not `NaN`, and is never flagged. `transform_tokens` does **not** zero-fill — an empty sequence stays all-`NaN` per token and is masked downstream in `BaseDetector.predict_token_proba` (#240).

### Warning class
`EmptySequenceWarning` is **project-specific**, not a sklearn warning (sklearn has no category for empty input; `UndefinedMetricWarning` is metrics-only and wrong here). Define a small `artefactual/exceptions.py`:
```python
class ArtefactualWarning(UserWarning):
    """Base for all artefactual warnings (subclass of UserWarning → shown by default, filterable)."""

class EmptySequenceWarning(ArtefactualWarning):
    """A sequence had no tokens; scored at the classifier baseline."""
```
Subclassing `UserWarning` keeps it visible by default but precisely filterable. This module is the shared home for the project's warnings — coordinate with #221 (parser "no top logprobs") and #225 (`NumericalPrecisionWarning`) so they live here too rather than as scattered bare `UserWarning`s.

## Class Sketch
```python
class EntropyTransformer(BaseEstimator, TransformerMixin, EntropyContributionMixin):
    def __init__(self, k: int = 15, reduction: str | callable = "epr"):
        self.k = k
        self.reduction = reduction          # "epr" | "wepr" | custom reduction(s_kj, axis)

    def __sklearn_tags__(self):
        tags = super().__sklearn_tags__()
        tags.requires_fit = False              # stateless: fit learns nothing from data
        tags.input_tags.allow_nan = True       # consumes NaN-padded input on purpose (else check_estimators_nan_inf fails)
        return tags

    def fit(self, X, y=None):
        return self                         # nothing to learn; transform works without fit

    @property
    def reduction_fn(self):
        # resolve string → function lazily (a callable passes through). NOT in __init__:
        # self.reduction must stay unchanged for clone/get_params/set_params.
        return self.reduction if callable(self.reduction) else STRATEGIES[self.reduction]

    def transform(self, X: np.ndarray) -> np.ndarray:          # (n, n_features)
        s_kj = self.entropy_contributions(X, self.k)
        features = self.reduction_fn(s_kj, axis=1)
        empty = np.isnan(features).all(axis=1)                  # token-less sequences → all-NaN row
        if empty.any():
            warnings.warn(                                      # surface the degenerate input, don't pass silently
                f"{int(empty.sum())} empty (token-less) sequence(s); scoring at baseline sigmoid(intercept_).",
                EmptySequenceWarning, stacklevel=2,
            )
        features[empty] = 0.0                                   # zero features → classifier returns baseline
        return features

    def transform_tokens(self, X: np.ndarray) -> np.ndarray:   # (n, T, n_features)
        s_kj = self.entropy_contributions(X, self.k)
        windows = np.expand_dims(s_kj, axis=2)                  # (n, T, 1, k) — one 1-token window per token
        return self.reduction_fn(windows, axis=2)
```

## Reduction function — args, input/output shapes
Signature: `reduction(s_kj: np.ndarray, axis: int) -> np.ndarray`.

**Arguments**
- `s_kj: np.ndarray` — NaN-aware entropy contributions from `EntropyContributionMixin` (#239). The **rank axis is always the last axis** (`-1`, size `k`); the **token axis is the one given by `axis`**. Real positions are finite, padding is `NaN`. Dtype float.
- `axis: int` — the token/window axis to pool over (the transformer passes `1` for sequence, `2` for token; see below). All token pooling must use this argument.

It is called with **two different shapes** — the token call inserts a length-1 window axis so the same function works at both levels:

| call | input `s_kj` | `axis` (token/window) | rank axis | output |
|---|---|---|---|---|
| `transform` (sequence) | `(n, T, k)` | `1` | `-1` | `(n, n_features)` |
| `transform_tokens` (token) | `(n, T, 1, k)` | `2` | `-1` | `(n, T, n_features)` |

Contract for a custom callable `reduction(s_kj, axis) -> ndarray`:
- The **rank axis is always the last axis** (`-1`), size `k`; do per-rank ops on `-1`.
- The **`axis` argument is the token axis to pool over** — use it for *all* token pooling; never hardcode `1` vs `2` or assume 3-D vs 4-D input.
- Pooling must be the **identity over a length-1 axis** (`nanmean`/`nanmax`/`nansum` are) — this is what makes the token call (a 1-token window) reproduce the per-token features and guarantees token/sequence consistency.
- Output drops the `axis` dimension; the last axis becomes `n_features` (`nansum`→`1`, `concat([mean, max])`→`2k`, …).

(`"wepr"` is registered in #242.)

## Acceptance criteria
- [ ] `get_params()`, `set_params()`, and `clone()` work (constructor stores `k` and `reduction` unchanged).
- [ ] Stateless: `__sklearn_tags__` sets `requires_fit = False`; `transform` / `transform_tokens` work **without** calling `fit`, and `fit` learns nothing from data.
- [ ] `__sklearn_tags__` sets `input_tags.allow_nan = True` (the transformer consumes NaN-padded input) — so `check_estimator` skips `check_estimators_nan_inf` instead of failing it (#245).
- [ ] `reduction` accepts `"epr"` (built-in `STRATEGIES`) or a custom `reduction(s_kj, axis)` callable.
- [ ] `transform(X)` returns `(n, n_features)`; `transform_tokens(X)` returns `(n, T, n_features)` — both from the *same* reduction function.
- [ ] EPR padded-token handling restores `NaN` after `nansum`, reproducing legacy EPR scores (mean over real tokens).
- [ ] `transform` detects empty (all-`NaN`) sequences, **emits `EmptySequenceWarning`** (project warning in `artefactual/exceptions.py`, subclass of `UserWarning`), and maps them to a zero feature vector → baseline `sigmoid(intercept_)`; `transform_tokens` leaves empty sequences all-`NaN` (masked in #240).

### PR DESCRIPTION
**Pull request by CharlotteLeBihan18**
closes #277 and #242

### CODE REVIEW

File: src/artefactual/scoring/entropy_methods/entropy_transformer.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import warnings
+from collections.abc import Callable
+
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils import Tags
+
+from artefactual.exceptions import EmptySequenceWarning
+from artefactual.scoring.entropy_methods.entropy_contributions import EntropyContributionsMixin
```
[the remark this answers is not in the document]
Comment (CharlotteLeBihan18) [reply src/artefactual/scoring/entropy_methods/entropy_transformer.py:9]: This is part of a stacked PR chain. EntropyContributionsMixin is introduced in PR #289 (issue-239-EntropyContributionMixin). Once the entire stack is merged sequentially, this import will resolve.

File: src/artefactual/scoring/entropy_methods/entropy_transformer.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    @property
+    def reduction_fn(self) -> Callable:
+        # self.reduction must stay unchanged for clone/get_params/set_params.
+        return self.reduction if callable(self.reduction) else STRATEGIES[self.reduction]
+
+    def transform(self, x: np.ndarray) -> np.ndarray:  # (n, n_features)
+        s_kj = self.entropy_contributions(x)
+        features = self.reduction_fn(s_kj, axis=1)
+        empty = np.isnan(features).all(axis=1)  # token-less sequences → all-NaN row
```
[the remark this answers is not in the document]
Comment (CharlotteLeBihan18) [reply src/artefactual/scoring/entropy_methods/entropy_transformer.py:56]: Both _epr and _wepr preserve the batch dimension, so axis=1 and axis=-1 are equivalent.
